src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs

```diff
@@ -0,0 +1,768 @@
+            {
+                throw new ArgumentOutOfRangeException(nameof(capacity));
+            }
+
+            int newSize = HashHelpers.GetPrime(capacity);
+            if (newSize < _entries.Length)
+            {
+                Resize(newSize);
+                ++_version;
+            }
+        }
+
+        public bool TryAdd(TKey key, TValue value) => TryInsert(null, key, value, InsertionBehavior.None);
+
+        public bool TryGetValue(TKey key, out TValue value)
+        {
+            int index = IndexOf(key);
+            if (index >= 0)
+            {
+                value = _entries[index].Value;
+                return true;
+            }
+            value = default;
+            return false;
+        }
+
+        #region Explicit Interface Implementation
+        KeyValuePair<TKey, TValue> IList<KeyValuePair<TKey, TValue>>.this[int index]
+        {
+            get
+            {
+                TValue value = GetAt(index, out TKey key);
+                return new KeyValuePair<TKey, TValue>(key, value);
+            }
+            set => throw new NotSupportedException(); // Should this be supported?
+        }
+
+        KeyValuePair<TKey, TValue> IReadOnlyList<KeyValuePair<TKey, TValue>>.this[int index] => ((IList<KeyValuePair<TKey, TValue>>)this)[index];
+
+        ICollection<TKey> IDictionary<TKey, TValue>.Keys => Keys;
+
+        ICollection<TValue> IDictionary<TKey, TValue>.Values => Values;
+
+        IEnumerable<TKey> IReadOnlyDictionary<TKey, TValue>.Keys => Keys;
+
+        IEnumerable<TValue> IReadOnlyDictionary<TKey, TValue>.Values => Values;
+
+        bool ICollection<KeyValuePair<TKey, TValue>>.IsReadOnly => false;
+
+        IEnumerator<KeyValuePair<TKey, TValue>> IEnumerable<KeyValuePair<TKey, TValue>>.GetEnumerator() => GetEnumerator();
+
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+        void ICollection<KeyValuePair<TKey, TValue>>.Add(KeyValuePair<TKey, TValue> item) => Add(item.Key, item.Value);
+
+        bool ICollection<KeyValuePair<TKey, TValue>>.Contains(KeyValuePair<TKey, TValue> item) => TryGetValue(item.Key, out TValue value) && EqualityComparer<TValue>.Default.Equals(value, item.Value);
+
+        void ICollection<KeyValuePair<TKey, TValue>>.CopyTo(KeyValuePair<TKey, TValue>[] array, int arrayIndex)
+        {
+            if (array == null)
+            {
+                throw new ArgumentNullException(nameof(array));
+            }
+            if (arrayIndex < 0 || arrayIndex > array.Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(arrayIndex));
+            }
+            int count = Count;
+            if (array.Length - arrayIndex < count)
+            {
+                throw new ArgumentException("Destination array is not long enough to copy all the items in the collection. Check array index and length.");
+            }
+
+            Entry[] entries = _entries;
+            for (int i = 0; i < count; ++i)
```

Comment:
Is it guaranteed that `count` is within `entries.Length`?

Reply:
Yes, the `Count` property is guaranteed to always be less than or equal to `_entries.Length`.

Reply:
Great. Debug.Assert?

Reply:
A Debug.Assert would not add anything because it would throw on the next line.